File: crates/tauri-plugin-gattify/android/src/main/java/dev/gattify/plugin/Wire.kt
package dev.gattify.plugin

import app.tauri.plugin.JSObject
import org.json.JSONArray
import org.json.JSONObject
import java.util.Base64

internal fun encodeBase64(bytes: ByteArray): String = Base64.getEncoder().encodeToString(bytes)

internal fun decodeBase64(value: String, field: String): ByteArray = try {
  Base64.getDecoder().decode(value)
} catch (_: IllegalArgumentException) {
  throw invalidArgument("$field is not valid base64")
}

/** Puts [value], or JSON null. `JSONObject.put` with a Kotlin null removes the key instead. */
private fun JSObject.putNullable(key: String, value: Any?): JSObject = put(key, value ?: JSONObject.NULL)

/** A scan result, in plain values so that the JSON builder runs without Android classes. */
internal class DeviceReport(
  val id: String,
  val name: String?,
  val rssi: Int?,
  val serviceUuids: List<String>,
  val localName: String?,
  val serviceData: List<Pair<String, ByteArray>>,
  val manufacturerData: List<Pair<Int, ByteArray>>,
  val connectable: Boolean?,
  val observedAtMillis: Long,
  val scanId: String,
)

internal class CharacteristicReport(val handle: String, val uuid: String, val properties: Properties)

internal class ServiceReport(val handle: String, val uuid: String, val characteristics: List<CharacteristicReport>)

internal class PermissionOutcomes(val scan: String, val connect: String, val advertise: String)

/** The facts that the capability report depends on. */
internal class CapabilityFacts(val hasAdapter: Boolean, val adapterOn: Boolean, val hasAdvertiser: Boolean)

/** The replies of `execute`: `{ kind }` or `{ kind, payload }`. */
internal object Replies {
  fun empty(): JSObject = JSObject().put("kind", "empty")

  private fun reply(kind: String, payload: Any): JSObject = JSObject().put("kind", kind).put("payload", payload)

  fun state(state: String) = reply("state", state)

  fun capabilities(facts: CapabilityFacts): JSObject {
    val central = if (facts.hasAdapter) "supported" to "available" else "unsupported" to "noAdapter"
    val peripheral = when {
      !facts.hasAdapter -> "unsupported" to "noAdapter"
      // The advertiser getter returns null while the adapter is off, so support is unknown then.
      !facts.adapterOn -> "unknown" to "bluetoothOff"
      facts.hasAdvertiser -> "supported" to "available"
      else -> "unsupported" to "noAdvertiser"
    }
    val payload = JSObject().put("central", support(central))
    for (key in listOf("peripheral", "advertising", "targetedNotify", "simultaneousRoles")) {
      payload.put(key, support(peripheral))
    }
    payload.put("background", support("unsupported" to "foregroundOnlyContract"))
    payload.putNullable("maxConnections", null)
    payload.putNullable("maxAdvertisingDataLength", if (facts.hasAdapter) LEGACY_ADVERTISING_DATA_LENGTH else null)
    return reply("capabilities", payload)
  }

  /** A `Support` from a level and a reason. */
  private fun support(levelAndReason: Pair<String, String>): JSObject = JSObject()
    .put("level", levelAndReason.first)
    .put("reason", levelAndReason.second)
    .putNullable("description", null)

  fun permissions(outcomes: PermissionOutcomes) = reply(
    "permissions",
    JSObject().put("scan", outcomes.scan).put("connect", outcomes.connect).put("advertise", outcomes.advertise),
  )

  fun scanStarted(scanId: String) = reply("scanStarted", JSObject().put("scanId", scanId))

  fun connected(connectionId: String, mtu: Int) =
    reply("connected", JSObject().put("connectionId", connectionId).put("limits", linkLimits(mtu)))

  fun linkLimits(mtu: Int): JSObject {
    val length = valueLength(mtu)
    return JSObject()
      .put("writeWithResponse", length)
      .put("writeWithoutResponse", length)
      .put("notification", length)
      .put("attMtu", mtu)
  }

  fun services(services: List<ServiceReport>): JSObject {
    val list = JSONArray()
    for (service in services) {
      val characteristics = JSONArray()
      for (characteristic in service.characteristics) {
        characteristics.put(
          JSObject()
            .put("handle", characteristic.handle)
            .put("uuid", characteristic.uuid)
            .put("properties", characteristic.properties.toJson()),
        )
      }
      list.put(
        JSObject().put("handle", service.handle).put("uuid", service.uuid).put("characteristics", characteristics),
      )
    }
    return reply("services", list)
  }

  fun bytes(value: ByteArray) = reply("bytes", JSObject().put("valueBase64", encodeBase64(value)))

  fun subscriptionStarted(subscriptionId: String) =
    reply("subscriptionStarted", JSObject().put("subscriptionId", subscriptionId))

  fun serverCreated(serverId: String) = reply("serverCreated", JSObject().put("serverId", serverId))

  fun advertisingStarted(localNameIncluded: Boolean, localNameTruncated: Boolean) = reply(
    "advertisingStarted",
    JSObject().put("localNameIncluded", localNameIncluded).put("localNameTruncated", localNameTruncated),
  )

  fun resources(scans: Int, connections: Int, subscriptions: Int, servers: Int) = reply(
    "resources",
    JSObject().put("scans", scans).put("connections", connections).put("subscriptions", subscriptions)
      .put("servers", servers),
  )
}

/** The events of the channel: `{ kind, payload }`, sent inside `{ ownerId, event }`. */
internal object Events {
  fun envelope(ownerId: String, event: JSObject): JSObject = JSObject().put("ownerId", ownerId).put("event", event)

  private fun event(kind: String, payload: JSObject): JSObject = JSObject().put("kind", kind).put("payload", payload)

  fun adapterStateChanged(state: String) = event("adapterStateChanged", JSObject().put("state", state))

  fun scanResult(device: DeviceReport): JSObject {
    val serviceData = JSONArray()
    for ((uuid, bytes) in device.serviceData) {
      serviceData.put(JSObject().put("serviceUuid", uuid).put("bytesBase64", encodeBase64(bytes)))
    }
    val manufacturerData = JSONArray()
    for ((companyId, bytes) in device.manufacturerData) {
      manufacturerData.put(JSObject().put("companyId", companyId).put("bytesBase64", encodeBase64(bytes)))
    }
    val advertisement = JSObject()
      .putNullable("localName", device.localName)
      .put("serviceData", serviceData)
      .put("manufacturerData", manufacturerData)
      .putNullable("connectable", device.connectable)
    val payload = JSObject()
      .put("id", device.id)
      .putNullable("name", device.name)
      .putNullable("rssi", device.rssi)
      .put("serviceUuids", JSONArray(device.serviceUuids))
      .put("advertisement", advertisement)
      .put("observedAtMillis", device.observedAtMillis)
      .put("scanId", device.scanId)
    return event("scanResult", JSObject().put("device", payload))
  }

  fun scanStopped(scanId: String) = event("scanStopped", JSObject().put("scanId", scanId))

  fun connectionClosed(connectionId: String) = event("connectionClosed", JSObject().put("connectionId", connectionId))

  fun characteristicValue(subscriptionId: String, value: ByteArray) = event(
    "characteristicValue",
    JSObject().put("subscriptionId", subscriptionId).put("valueBase64", encodeBase64(value)),
  )

  fun serverWrite(serverId: String, peerId: String, characteristicKey: String, value: ByteArray) = event(
    "serverWrite",
    JSObject()
      .put("serverId", serverId)
      .put("peerId", peerId)
      .put("characteristicKey", characteristicKey)
      .put("valueBase64", encodeBase64(value)),
  )

  fun subscriptionChanged(serverId: String, peerId: String, characteristicKey: String, maxValueLength: Int?) = event(
    "subscriptionChanged",
    JSObject()
      .put("serverId", serverId)
      .put("peerId", peerId)
      .put("characteristicKey", characteristicKey)
      .put("subscribed", maxValueLength != null)
      .putNullable("maxValueLength", maxValueLength),
  )

  fun criticalStateLoss(resourceId: String, reason: String) =
    event("criticalStateLoss", JSObject().put("resourceId", resourceId).put("reason", reason))
}

/** A legacy advertisement holds 31 bytes. */
internal const val LEGACY_ADVERTISING_DATA_LENGTH = 31

File: crates/tauri-plugin-gattify/ios/Sources/Wire.swift
import Foundation

/// The longest attribute value in ATT.
let maxAttributeValueLength = 512

/// The foreground advertising bytes iOS gives an app.
let maxAdvertisingDataLength = 28

func clampedValueLength(_ length: Int) -> Int {
  min(length, maxAttributeValueLength)
}

struct LinkLimits: Equatable {
  let writeWithResponse: Int
  let writeWithoutResponse: Int
  let notification: Int
  let attMtu: Int

  /// iOS reports one value length for the link: the ATT MTU minus 3.
  init(maximumWriteValueLength length: Int) {
    let value = clampedValueLength(length)
    writeWithResponse = value
    writeWithoutResponse = value
    notification = value
    attMtu = length + 3
  }

  var json: JSON {
    [
      "writeWithResponse": .int(writeWithResponse),
      "writeWithoutResponse": .int(writeWithoutResponse),
      "notification": .int(notification),
      "attMtu": .int(attMtu),
    ]
  }
}

struct CharacteristicInstance: Equatable {
  let handle: String
  let uuid: String
  let properties: CharacteristicProperties
}

struct ServiceInstance: Equatable {
  let handle: String
  let uuid: String
  let characteristics: [CharacteristicInstance]
}

struct ResourceCounts: Equatable {
  var scans = 0
  var connections = 0
  var subscriptions = 0
  var servers = 0
}

enum Reply: Equatable {
  case empty
  case state(String)
  case capabilities
  case permissions(String)
  case scanStarted(scanId: String)
  case connected(connectionId: String, limits: LinkLimits)
  case services([ServiceInstance])
  case bytes(Data)
  case subscriptionStarted(subscriptionId: String)
  case serverCreated(serverId: String)
  case advertisingStarted(localNameIncluded: Bool, localNameTruncated: Bool)
  case resources(ResourceCounts)

  var json: JSON {
    switch self {
    case .empty:
      return ["kind": "empty"]
    case .state(let state):
      return ["kind": "state", "payload": .string(state)]
    case .capabilities:
      return ["kind": "capabilities", "payload": capabilitiesPayload]
    case .permissions(let outcome):
      let value = JSON.string(outcome)
      return ["kind": "permissions", "payload": ["scan": value, "connect": value, "advertise": value]]
    case .scanStarted(let scanId):
      return ["kind": "scanStarted", "payload": ["scanId": .string(scanId)]]
    case .connected(let connectionId, let limits):
      return [
        "kind": "connected",
        "payload": ["connectionId": .string(connectionId), "limits": limits.json],
      ]
    case .services(let services):
      return ["kind": "services", "payload": .array(services.map(\.json))]
    case .bytes(let value):
      return ["kind": "bytes", "payload": ["valueBase64": .string(value.base64EncodedString())]]
    case .subscriptionStarted(let subscriptionId):
      return ["kind": "subscriptionStarted", "payload": ["subscriptionId": .string(subscriptionId)]]
    case .serverCreated(let serverId):
      return ["kind": "serverCreated", "payload": ["serverId": .string(serverId)]]
    case .advertisingStarted(let included, let truncated):
      return [
        "kind": "advertisingStarted",
        "payload": ["localNameIncluded": .bool(included), "localNameTruncated": .bool(truncated)],
      ]
    case .resources(let counts):
      return [
        "kind": "resources",
        "payload": [
          "scans": .int(counts.scans),
          "connections": .int(counts.connections),
          "subscriptions": .int(counts.subscriptions),
          "servers": .int(counts.servers),
        ],
      ]
    }
  }
}

private let capabilitiesPayload: JSON = {
  let supported: JSON = ["level": "supported", "reason": "available", "description": nil]
  return [
    "central": supported,
    "peripheral": supported,
    "advertising": supported,
    "targetedNotify": supported,
    "simultaneousRoles": supported,
    "background": ["level": "unsupported", "reason": "foregroundOnlyContract", "description": nil],
    "maxConnections": nil,
    "maxAdvertisingDataLength": .int(maxAdvertisingDataLength),
  ]
}()

struct ServiceDataEntry: Equatable {
  let uuid: String
  let bytes: Data
}

struct ManufacturerDataEntry: Equatable {
  let companyId: Int
  let bytes: Data

  /// Splits advertised manufacturer data into its little-endian company ID and the rest.
  init?(_ data: Data) {
    let bytes = [UInt8](data)
    guard bytes.count >= 2 else { return nil }
    companyId = Int(bytes[0]) | Int(bytes[1]) << 8
    self.bytes = Data(bytes[2...])
  }

  init(companyId: Int, bytes: Data) {
    self.companyId = companyId
    self.bytes = bytes
  }
}

struct AdvertisementData: Equatable {
  var localName: String?
  var serviceData: [ServiceDataEntry] = []
  var manufacturerData: [ManufacturerDataEntry] = []
  var connectable: Bool?

  var json: JSON {
    [
      "localName": JSON(localName),
      "serviceData": .array(
        serviceData.map { entry in
          ["serviceUuid": .string(entry.uuid), "bytesBase64": .string(entry.bytes.base64EncodedString())]
        }),
      "manufacturerData": .array(
        manufacturerData.map { entry in
          ["companyId": .int(entry.companyId), "bytesBase64": .string(entry.bytes.base64EncodedString())]
        }),
      "connectable": JSON(connectable),
    ]
  }
}

struct DiscoveredDevice: Equatable {
  let id: String
  let name: String?
  let rssi: Int?
  let serviceUuids: [String]
  let advertisement: AdvertisementData
  let observedAtMillis: Int
  let scanId: String

  var json: JSON {
    [
      "id": .string(id),
      "name": JSON(name),
      "rssi": JSON(rssi),
      "serviceUuids": .array(serviceUuids.map(JSON.string)),
      "advertisement": advertisement.json,
      "observedAtMillis": .int(observedAtMillis),
      "scanId": .string(scanId),
    ]
  }
}

enum BridgeEvent: Equatable {
  case adapterStateChanged(state: String)
  case scanResult(DiscoveredDevice)
  case scanStopped(scanId: String)
  case connectionClosed(connectionId: String)
  case characteristicValue(subscriptionId: String, value: Data)
  case serverWrite(serverId: String, peerId: String, characteristicKey: String, value: Data)
  case subscriptionChanged(
    serverId: String, peerId: String, characteristicKey: String, subscribed: Bool,
    maxValueLength: Int?)
  case criticalStateLoss(resourceId: String, reason: String)

  var kind: String {
    switch self {
    case .adapterStateChanged: return "adapterStateChanged"
    case .scanResult: return "scanResult"
    case .scanStopped: return "scanStopped"
    case .connectionClosed: return "connectionClosed"
    case .characteristicValue: return "characteristicValue"
    case .serverWrite: return "serverWrite"
    case .subscriptionChanged: return "subscriptionChanged"
    case .criticalStateLoss: return "criticalStateLoss"
    }
  }

  var payload: JSON {
    switch self {
    case .adapterStateChanged(let state):
      return ["state": .string(state)]
    case .scanResult(let device):
      return ["device": device.json]
    case .scanStopped(let scanId):
      return ["scanId": .string(scanId)]
    case .connectionClosed(let connectionId):
      return ["connectionId": .string(connectionId)]
    case .characteristicValue(let subscriptionId, let value):
      return ["subscriptionId": .string(subscriptionId), "valueBase64": .string(value.base64EncodedString())]
    case .serverWrite(let serverId, let peerId, let key, let value):
      return [
        "serverId": .string(serverId),
        "peerId": .string(peerId),
        "characteristicKey": .string(key),
        "valueBase64": .string(value.base64EncodedString()),
      ]
    case .subscriptionChanged(let serverId, let peerId, let key, let subscribed, let maxValueLength):
      return [
        "serverId": .string(serverId),
        "peerId": .string(peerId),
        "characteristicKey": .string(key),
        "subscribed": .bool(subscribed),
        "maxValueLength": JSON(maxValueLength),
      ]
    case .criticalStateLoss(let resourceId, let reason):
      return ["resourceId": .string(resourceId), "reason": .string(reason)]
    }
  }

  func envelope(ownerId: String) -> JSON {
    ["ownerId": .string(ownerId), "event": ["kind": .string(kind), "payload": payload]]
  }
}

extension CharacteristicProperties {
  var json: JSON {
    [
      "read": .bool(read),
      "write": .bool(write),
      "writeWithoutResponse": .bool(writeWithoutResponse),
      "notify": .bool(notify),
      "indicate": .bool(indicate),
    ]
  }
}

extension CharacteristicInstance {
  var json: JSON {
    ["handle": .string(handle), "uuid": .string(uuid), "properties": properties.json]
  }
}

extension ServiceInstance {
  var json: JSON {
    ["handle": .string(handle), "uuid": .string(uuid), "characteristics": .array(characteristics.map(\.json))]
  }
}
